src/Control/Lens/Type.hs

```diff
@@ -622,3 +626,53 @@ type Over p f s t a b = p a (f b) -> s -> f t
 -- type 'Over'' p f = 'Simple' ('Over' p f)
 -- @
 type Over' p f s a = Over p f s s a a
+
+-- | Extract the concrete 'Functor' from a concrete 'ALens', 'AnIso', etc.
+--
+-- @
+-- GetConcreteFunctor (ALens s t a b) = Pretext (->) a b
+-- GetConcreteFunctor (AnIso s t a b) = Identity
+-- @
+#if __GLASGOW_HASKELL__ >= 800
+type family GetConcreteFunctor (x :: *) :: GetConcreteFunctorKind x
+type instance GetConcreteFunctor x = GetConcreteFunctorType (GetConcreteFunctorKind x) x
+type family GetConcreteFunctorKind (x :: *) :: *
+type instance GetConcreteFunctorKind (pafb -> p s ((f :: j -> k) t)) = j -> k
+type family GetConcreteFunctorType (k :: *) (x :: *) :: k
+type instance GetConcreteFunctorType (i -> o) (pafb -> p s ((f :: i -> o) t)) = f
+#else
+type family GetConcreteFunctor (x :: *) :: * -> *
+type instance GetConcreteFunctor (pafb -> (p :: * -> * -> *) s ((f :: * -> *) t)) = f
+#endif
+
+-- | Extract the concrete 'Profunctor' from a concrete 'ALens', 'AnIso', etc.
```

Comment:
You could delegate to another type family recursively to handle (p s (f t)) which delegates to another to force p to have kind * -> * -> *. This would likely require ~3 type families though.